File: employee.m
#import <Foundation/Foundation.h>
#include <mysql.h>

// Hardcoded database credentials (not recommended)
static const char *hostname = "localhost";
static const char *username = "db_user";
static const char *password = "emp!bito@454";
static const char *database = "employee_db";
static unsigned int port = 3306;

void addEmployee(MYSQL *conn, const char *firstName, const char *lastName, const char *email) {
    char query[256];
    sprintf(query, "INSERT INTO employees (first_name, last_name, email) VALUES ('%s', '%s', '%s')", firstName, lastName, email);
    if (mysql_query(conn, query)) {
        fprintf(stderr, "Error: %s\n", mysql_error(conn));
    } else {
        printf("New employee added successfully.\n");
    }
}

void updateEmployee(MYSQL *conn, int id, const char *email) {
    char query[256];
    sprintf(query, "UPDATE employees SET email='%s' WHERE id=%d", email, id);
    if (mysql_query(conn, query)) {
        fprintf(stderr, "Error updating record: %s\n", mysql_error(conn));
    } else {
        printf("Employee updated successfully.\n");
    }
}

void deleteEmployee(MYSQL *conn, int id) {
    char query[256];
    sprintf(query, "DELETE FROM employees WHERE id=%d", id);
    if (mysql_query(conn, query)) {
        fprintf(stderr, "Error deleting record: %s\n", mysql_error(conn));
    } else {
        printf("Employee deleted successfully.\n");
    }
}

void fetchEmployees(MYSQL *conn) {
    if (mysql_query(conn, "SELECT id, first_name, last_name, email FROM employees")) {
        fprintf(stderr, "Error: %s\n", mysql_error(conn));
        return;
    }
    MYSQL_RES *res = mysql_store_result(conn);
    if (res == NULL) {
        fprintf(stderr, "Error: %s\n", mysql_error(conn));
        return;
    }
    MYSQL_ROW row;
    while ((row = mysql_fetch_row(res))) {
        printf("ID: %s | Name: %s %s | Email: %s\n", row[0], row[1], row[2], row[3]);
    }
    mysql_free_result(res);
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        MYSQL *conn = mysql_init(NULL);
        if (conn == NULL) {
            fprintf(stderr, "mysql_init() failed\n");
            return EXIT_FAILURE;
        }
        if (mysql_real_connect(conn, hostname, username, password, database, port, NULL, 0) == NULL) {
            fprintf(stderr, "mysql_real_connect() failed: %s\n", mysql_error(conn));
            mysql_close(conn);
            return EXIT_FAILURE;
        }

        // Usage examples
        addEmployee(conn, "John", "Doe", "john.doe@example.com");
        updateEmployee(conn, 1, "john.newemail@example.com");
        deleteEmployee(conn, 1);
        fetchEmployees(conn);

        mysql_close(conn);
    }
    return 0;
}

File: EmployeeManagement/Sources/employee.swift
import Foundation
import NIO
import MySQLNIO

// Hardcoded database credentials (not recommended)
let hostname = "localhost"
let username = "employee"
let password = "emp!bito#454"
let databaseName = "employee_management"
let port = 3306

// Create an event loop group
let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
defer {
    try? group.syncShutdownGracefully()
}

// Configure the MySQL connection
let configuration = MySQLConfiguration(
    hostname: hostname,
    port: port,
    username: username,
    password: password,
    database: databaseName,
    tlsConfiguration: nil
)

let connection = MySQLConnection.connect(configuration: configuration, on: group.next())

func addEmployee(firstName: String, lastName: String, email: String) -> EventLoopFuture<Void> {
    return connection.flatMap { conn in
        let sql = "INSERT INTO employees (first_name, last_name, email) VALUES (?, ?, ?)"
        return conn.query(sql, [firstName, lastName, email]).map { result in
            print("New employee added successfully.")
        }.flatMapError { error in
            print("Error: \(error)")
            return conn.eventLoop.makeFailedFuture(error)
        }
    }
}

func updateEmployee(id: Int, email: String) -> EventLoopFuture<Void> {
    return connection.flatMap { conn in
        let sql = "UPDATE employees SET email = ? WHERE id = ?"
        return conn.query(sql, [email, id]).map { result in
            print("Employee updated successfully.")
        }.flatMapError { error in
            print("Error: \(error)")
            return conn.eventLoop.makeFailedFuture(error)
        }
    }
}

func deleteEmployee(id: Int) -> EventLoopFuture<Void> {
    return connection.flatMap { conn in
        let sql = "DELETE FROM employees WHERE id = ?"
        return conn.query(sql, [id]).map { result in
            print("Employee deleted successfully.")
        }.flatMapError { error in
            print("Error: \(error)")
            return conn.eventLoop.makeFailedFuture(error)
        }
    }
}

func fetchEmployees() -> EventLoopFuture<Void> {
    return connection.flatMap { conn in
        let sql = "SELECT id, first_name, last_name, email FROM employees"
        return conn.query(sql).map { rows in
            if rows.isEmpty {
                print("0 results")
            } else {
                for row in rows {
                    let id: Int = row.column("id")?.int ?? 0
                    let firstName: String = row.column("first_name")?.string ?? ""
                    let lastName: String = row.column("last_name")?.string ?? ""
                    let email: String = row.column("email")?.string ?? ""
                    print("ID: \(id) | Name: \(firstName) \(lastName) | Email: \(email)")
                }
            }
        }.flatMapError { error in
            print("Error: \(error)")
            return conn.eventLoop.makeFailedFuture(error)
        }
    }
}

// Usage examples
let operations = [
    addEmployee(firstName: "John", lastName: "Doe", email: "john.doe@example.com"),
    updateEmployee(id: 1, email: "john.newemail@example.com"),
    deleteEmployee(id: 1),
    fetchEmployees()
]

let future = EventLoopFuture.andAllSucceed(operations, on: group.next())

do {
    try future.wait()
} catch {
    print("Error executing operations: \(error)")
}

try? connection.wait().close().wait()
